### PR TITLE
feat(DTFS2-7489): create submit cancellation endpoint tfm api

### PR DESCRIPTION
**Pull request by MarRobSoftwire**
## Introduction :pencil2:
The user needs to be able to submit the cancellation. This endpoint provides the stub from which the database update & email can be sent from.

## Resolution :heavy_check_mark:
- Add submit cancellation endpoint with middleware validation
- Add api and unit test coverage

## Note ⚠️ 
This endpoint doesn't actually submit the cancellation yet, but merging this early will allow DTFS2-7298 and DTFS2-7364 to be completed in parallel

### CODE REVIEW

File: trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        description: 'the reason is undefined',
+        payload: { ...aValidPayload(), reason: undefined },
+      },
+      {
+        description: 'the bankRequestDate is a string ',
+        payload: { ...aValidPayload(), bankRequestDate: new Date().toISOString() },
+      },
+      {
+        description: 'the bankRequestDate is undefined ',
+        payload: { ...aValidPayload(), bankRequestDate: undefined },
```
Comment (ttbarnes) [inline trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts:120]: I think in such POST validation tests, instead of undefined, it should be an empty string `''`, since this is what `req.body` would contain if nothing is provided - same on line 112 and 128,

Comment (ttbarnes) [inline trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts:120]: I think in such POST validation tests, instead of undefined, it should be an empty string `''`, since this is what `req.body` would contain if nothing is provided - same on line 112 and 128,

**Review by ttbarnes — CHANGES_REQUESTED** [review]
Few minor comments, otherwise LGTM

File: trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts (hunk 1 above)
Comment (ttbarnes) [reply trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts:120]: Same in the middleware

File: trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      const url = getSubmitTfmDealCancellationUrl({ id: 'invalid' });
+
+      // Act
+      const response = await as(aPimUser).post(aValidPayload()).to(url);
+
+      // Assert
+      expect(response.status).toEqual(400);
+    });
+
+    const invalidPayloads = [
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts:101]: I think its nice to check the middleware is working as expected in an api test as well as the unit test - but you're right I can be less exhaustive. I'll just include a single `when the payload is invalid` test

File: trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts (hunk 1 above)
Comment (MarRobSoftwire) [reply trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts:120]: I dont think I quite understand what you mean
> since this is what req.body would contain if nothing is provided
Its possible for the payload to not include a property, so it would be undefined

File: trade-finance-manager-api/src/v1/controllers/deal-cancellation/submit-deal-cancellation.controller.test.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+const dealCancellation = {
+  reason: 'test reason',
+  bankRequestDate: 1794418807,
+  effectiveFrom: 1794418808,
+};
+
+const mockDealId = new ObjectId();
+const mockUserId = new ObjectId();
+
+describe('controllers - deal cancellation', () => {
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply trade-finance-manager-api/src/v1/controllers/deal-cancellation/submit-deal-cancellation.controller.test.ts:17]: I've got the differentiating describe within this `'POST - submitDealCancellation'` - I think its an existing pattern for the top level describe to refer to the group of controllers (in this case those serving endpoints `/deal/:id/cancellation`).

[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply trade-finance-manager-api/src/v1/controllers/deal-cancellation/submit-deal-cancellation.controller.test.ts:17]: I can refactor this (but should probably do the same to the put/get/delete tests too for consistency) if you like

[the remark this answers is not in the document]
Comment (ttbarnes) [reply trade-finance-manager-api/src/v1/controllers/deal-cancellation/submit-deal-cancellation.controller.test.ts:17]: If it's an existing pattern we should stick to it

File: trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts (hunk 1 above)
Comment (ttbarnes) [reply trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts]: A typical use case would be submitting an empty form, or the form with valid values apart from `bankRequestDate`. In this case it'd come through as `bankRequestDate: ''`

The only way AFAIK It would be `undefined` is if a user is manipulating the DOM or network request

I'd prefer this to be an empty string rather than specifically `undefined`. What do you think?

Comment (ttbarnes) [reply trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts]: Sorry @MarRobSoftwire I was thinking this was UI POST test 🤦 

Happy with this but should we add a test for if `bankRequestDate` is an empty string?

Comment (MarRobSoftwire) [reply trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts]: I've got a test
```
    {
      description: "the 'bankRequestDate' is a string",
      payload: {
        ...aValidPayload(),
        bankRequestDate: new Date().toString(),
      },
    },
```
The zod validation is looking for bankRequestDate to be a number so i dont think we gain anything new by checking for an empty string?

Comment (ttbarnes) [reply trade-finance-manager-api/api-tests/v1/deal-cancellation/submit-deal-cancellation.post.api-test.ts]: OK makes sense, thanks!

**Review by ttbarnes — APPROVED** [review]
